crates/exex/src/manager.rs

```diff
@@ -269,75 +260,114 @@ impl Future for ExExManager {
+
+            // ensure the exex has capacity
+            let mut reserve_result = exex.poll_reserve(cx);
+
+            // drain notifications
+            loop {
+                match reserve_result {
+                    Poll::Ready(Ok(())) => {
+                        // try send next notification to exex if we have it
+                        let notification_id = exex.next_notification_id;
+                        if let Some(notification) = self.buffer.get(notification_id) {
+                            debug!(exex.id, notification_id, "sent notification to exex");
+                            if let Poll::Ready(Err(err)) = exex.send(notification) {
+                                // the channel was closed, which is irrecoverable for the manager
+                                return Poll::Ready(Err(err.into()))
+                            }
+                        } else {
+                            // No more notifications to send
+                            break;
+                        }
+                    }
+                    Poll::Ready(Err(err)) => return Poll::Ready(Err(err.into())),
+                    Poll::Pending => break,
                 }
+
+                // update reserve_result for the next iteration
+                reserve_result = exex.poll_reserve(cx);
             }
-            min_id = min_id.min(exex.next_notification_id);
-            self.exex_handles.push(exex);
-        }
 
-        // remove processed buffered notifications
-        self.buffer.retain(|&(id, _)| id >= min_id);
-        self.min_id = min_id;
-        debug!(min_id, "lowest notification id in buffer updated");
-
-        // update capacity
-        self.update_capacity();
-
-        // handle incoming exex events
-        for exex in self.exex_handles.iter_mut() {
-            while let Poll::Ready(Some(event)) = exex.receiver.poll_recv(cx) {
-                debug!(?event, id = exex.id, "received event from exex");
-                exex.metrics.events_sent_total.increment(1);
-                match event {
-                    ExExEvent::FinishedHeight(height) => exex.finished_height = Some(height),
+
+                // it is a logic error for this to ever underflow since the manager manages the
+                // notification IDs
+                let notification_id = exex
+                    .next_notification_id
+                    .checked_sub(self.min_id)
+                    .expect("exex expected notification ID outside the manager's range");
+                if let Some(notification) = self.buffer.get(notification_id) {
+                    debug!(exex.id, notification_id, "sent notification to exex");
+                    if let Poll::Ready(Err(err)) = exex.send(notification) {
+                        // the channel was closed, which is irrecoverable for the manager
+                        return Poll::Ready(Err(err.into()))
+                    }
```

Comment:
I think this is no longer accurate here, sending should only happen after we ensured there's capacity

Reply:
yes, we should not send notifications in 2 places in the loop, just the first when a slot in the channel has been reserved is needed

Reply:
Now senfing when Poll:Ready(Ok()) is matched